caikit/runtime/client/remote_model_finder.py

```diff
@@ -195,110 +238,139 @@ def _safe_discover(self) -> Dict[str, str]:
-
-        target = f"{self._connection.hostname}:{self._connection.port}"
-        options = tuple(self._connection._options.items())
-        with construct_grpc_channel(target, options, self._tls) as channel:
-            info_service_rpc = channel.unary_unary(
-                get_grpc_route_name(ServiceType.INFO, "GetModelsInfo"),
-                request_serializer=ModelInfoRequest.get_proto_class().SerializeToString,
-                response_deserializer=ModelInfoResponse.get_proto_class().FromString,
-            )
-            try:
-                model_info_proto = info_service_rpc(
-                    ModelInfoRequest().to_proto(), timeout=self._connection.timeout
-                )
-            except grpc.RpcError as exc:
-                log.warning(
-                    "Unable to discover modules from remote: %s. Error: %s",
-                    self._connection.hostname,
-                    str(exc),
-                )
-                return {}
-
-        model_info_response = ModelInfoResponse.from_proto(model_info_proto)
-
-        # Parse response into dictionary of name->id
         supported_modules = {}
-        for model_info in model_info_response.models:
-            model_name = model_info.name
-            module_id = model_info.module_id
-
-            log.debug(
-                "Discovered model %s with module_id %s from remote runtime",
-                model_name,
-                module_id,
-            )
-            supported_modules[model_name] = module_id
+        for conn in self._get_conn_candidates(model_name):
+            target = f"{conn.hostname}:{conn.port}"
+            options = [tuple(opt) for opt in conn.options.items()]
+            with construct_grpc_channel(target, options, conn.tls) as channel:
+                info_service_rpc = channel.unary_unary(
+                    get_grpc_route_name(ServiceType.INFO, "GetModelsInfo"),
+                    request_serializer=ModelInfoRequest.get_proto_class().SerializeToString,
+                    response_deserializer=ModelInfoResponse.get_proto_class().FromString,
+                )
+                try:
+                    model_info_proto = info_service_rpc(
+                        ModelInfoRequest().to_proto(), timeout=conn.timeout
+                    )
+
+                    model_info_response = ModelInfoResponse.from_proto(model_info_proto)
+
+                    # Parse response into dictionary of name->conn
+                    for model_info in model_info_response.models:
+                        model_name = model_info.name
+                        module_id = model_info.module_id
+
+                        log.debug(
+                            "Discovered model %s with module_id %s from remote runtime %s",
+                            model_name,
+                            module_id,
+                            target,
+                        )
+                        supported_modules[model_name] = ModuleConnectionInfo(
```

Comment:
since we are saving / caching the connection here, would this create problem for the cases where the pods gets refreshed and so the old connection object is now invalid?

Reply:
Yeah, this is definitely a key observation: There are several assumptions about the service behind the connection:

1. It is persistent. This is most easily accomplished by having the connection connect to a `Service` instead of `Pod`, but that then leads to (2)
2. It is homogeneous such that the model info endpoint will respond with the same set of loaded models, regardless of which replica is hit behind a service router.